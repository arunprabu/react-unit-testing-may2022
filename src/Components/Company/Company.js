import React, { useState } from 'react'

const Company = (props) => {

  const [ myCountry, setMyCountry] = useState('');
  
  const handleCountryNameChange = (event) => {
    console.log(event.target.value);
    setMyCountry(event.target.value);
  }

  return (
    <div>
      <h2>Company | Testing Props, Events, States, Hooks, JSX, Snapshot</h2>
      <h3>Welcome to My Company's Official Website!</h3>
      <p data-testid="companyNameEl">{props.name}</p>

      <div data-testid="employeesCountEl" 
        style={{padding: '10px', 
        color: 'green'}}>
        Employees: 318000+
      </div>

      {/* TODO: Have JOIN NOW button for Test Spec #5*/}

      <input type='text' placeholder='Enter Country' 
        value={myCountry} onChange={handleCountryNameChange}/>
      <p>Head over to Cognizant {myCountry} Website!</p>

    </div>
  )
}

export default Company