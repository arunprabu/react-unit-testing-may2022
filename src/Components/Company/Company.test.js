import Company from './Company';
import { render, screen, fireEvent } from '@testing-library/react';

describe('Company Component', () => { 

  // Test spec #1 
  it('should be available with in app', () => {
    expect(Company).toBeDefined();
  });

  // Test Spec #2 - Let's test the props
  it('receives props and displays them properly', () => {
    // take the comp to put under test with props
    render(<Company name='Cognizant Technologies Pvt Ltd'/>);
    // fetching the element by an attribute data-testid from JSX - Refer Company.js 
    // checking if the props are received correctly
    expect(screen.getByTestId('companyNameEl')).toHaveTextContent('Cognizant Technologies Pvt Ltd');
  });

  // Test Spec #3 
  it('has Welcome to My Company\'s Official Website!', ()=>{
    render(<Company />);
    const officialTxtEl = screen.getByText('Welcome to My Company\'s Official Website!');
    expect(officialTxtEl).toBeInTheDocument();
  });

  // Test spec #4
  it('has css style with color green', () => {
    render(<Company />);
    expect(screen.getByTestId('employeesCountEl')).toHaveStyle('color: green');
  });   

  // Test Spec #5
  // TODO: test whether the el has css class btn.
  // Do not use getByTestId. instead use getByRole
  // Have the button with label JOIN NOW 
  it('has a button with css class btn', () => {
    // have the assertion 
  });

  // Test Spec #6
  it(`has an element with placeholder 'Enter Country'`, () => {
    render(<Company />);
    const countryInput = screen.getByPlaceholderText('Enter Country');
    expect(countryInput).toBeTruthy();
  });

  // Test #7 - checking if the input element works with onChange event 
  it('should update country on onChange event', () => {
    render(<Company />);
    // find an element with placeholder Enter Country
    const countryInput = screen.getByPlaceholderText('Enter Country');
    // now checking if onChange is working or not
    // mock fire change event with the mock value 
    fireEvent.change(countryInput, { target: { value: 'Canada'} });
    // finally checking if the input element is showing the right country
    expect(countryInput.value).toBe('Canada');
  });

});