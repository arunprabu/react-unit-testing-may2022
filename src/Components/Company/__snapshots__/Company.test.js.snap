// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Company Component should have right company snapshot 1`] = `
<div>
  <h2>
    Company | Testing Props, Events, States, Hooks, JSX and Snapshot 
  </h2>
  <h3>
    Welcome to My Company's Official Website!
  </h3>
  <p
    data-testid="companyNameEl"
  >
    Cognizant Tech
  </p>
  <div
    data-testid="employeesCountEl"
    style={
      Object {
        "color": "green",
        "padding": "10px",
      }
    }
  >
    Employees: 318000+
  </div>
  <button
    className="btn btn-primary"
    role="button"
  >
    Join Cognizant
  </button>
  <br />
  <input
    onChange={[Function]}
    placeholder="Enter Country"
    type="text"
    value=""
  />
  <p>
    Head over to Cognizant 
    
     Website!
  </p>
</div>
`;
